src/display/network_utils.js

```diff
@@ -56,15 +55,199 @@ function validateRangeRequestCapabilities({ getResponseHeader, isHttp,
 function extractFilenameFromHeader(getResponseHeader) {
   const contentDisposition = getResponseHeader('Content-Disposition');
   if (contentDisposition) {
-    let parts =
-      /.+;\s*filename=(?:'|")(.+\.pdf)(?:'|")/gi.exec(contentDisposition);
-    if (parts !== null && parts.length > 1) {
-      return getFilenameFromUrl(parts[1]);
+    let filename = getFilenameFromContentDispositionHeader(contentDisposition);
+    if (/\.pdf$/i.test(filename)) {
+      return filename;
     }
   }
   return null;
 }
 
+// The logic in getFilenameFromContentDispositionHeader is adapted from
+// https://github.com/Rob--W/open-in-browser/blob/9f5fcae11cf6d99c503a15894f22efdfcd2075b7/extension/content-disposition.js
+// with the following changes:
+// - modified to conform to PDF.js's coding style.
+// - Support UTF-8 decoding when TextDecoder is unsupported.
+// - Replace "startsWith" with other string method for compat with old browsers.
+// - Move return to the end of the function to prevent Babel from dropping the
+//   function declarations.
+function getFilenameFromContentDispositionHeader(contentDisposition) {
```

Comment:
Since this function is basically "external" code, similar to other imported code (e.g. streams/webL10n), can't we just place it in the [external/ directory](https://github.com/mozilla/pdf.js/tree/master/external) instead, e.g. `external/content-disposition/` or similar?

Reply:
The `external/` directory doesn't seem to be used for direct inclusion of code, except for the `streams` library. Even in this case, there is extra code in `gulpfile.js` to manage the inclusion.

If you prefer a separate file, I can put it in a new file content-disposition.js.

The helper functions need to be in the scope of the main function, because the `needsEncodingFixup` variable needs to be updated:
https://github.com/Rob--W/pdf.js/blob/a59a81ad1fe0041f4cb45b4477079193f917b190/src/display/network_utils.js#L75
If I had to refactor the code to move the maintenance of this variable to a separate function parameter and/or return value, the code would look more messier. I would then have to pass an extra parameter to `textdecode` (and all of its callers), and either update this parameter (i.e. box the boolean in an object), or change the return value to return a pair (the boolean and the original return value).

Given this explanation, do you want any more changes before merging?

Reply:
> The helper functions need to be in the scope of the main function, because the needsEncodingFixup variable needs to be updated:

I overlooked that, sorry!

> If you prefer a separate file, I can put it in a new file content-disposition.js.
> [...]
> Given this explanation, do you want any more changes before merging?

I do like this idea, so please let's move the new code (verbatim) to a new `src/display/content-disposition.js` file; and then run the full test-suite before landing.

Reply:
Done. `gulp unittestcli` passes locally.

`unittestcli` is also run by Travis, so if the latest comit has a green checkmark, you can merge it.

Reply:
Couldn't the "Improve correctness of Content-Disposition parser" and "Move content disposition logic to separate file" commits be merged, to reduce the churn in the affected code a bit?

Reply:
Done.